libs/ml/src/ops/dataholder.cpp

```diff
@@ -83,15 +88,16 @@ bool DataHolder<TensorType>::SetData(TensorType const &data)
 
 template <class TensorType>
 std::vector<math::SizeType> DataHolder<TensorType>::ComputeOutputShape(
-    VecTensorType const &inputs) const
+    std::vector<math::SizeVector> const &inputs) const
 {
   FETCH_UNUSED(inputs);
-  if (!data_)
+  if (!future_data_shape_.empty())
   {
-    throw std::runtime_error("Data is not set for " + std ::string(Descriptor()) +
-                             ", shape computing is not possible.");
+    return future_data_shape_;
   }
-  return data_->shape();
+
+  throw std::runtime_error("future_data_shape_ is not set for " + std ::string(Descriptor()) +
```

Comment:
maybe worth flipping the return and throw positions here, so instead check for:
if (future_data_shape_.empty())